### PR TITLE
feat(api): add CORS headers

### PR DESCRIPTION
**Pull request by ojeytonwilliams**
Checklist:

- [x] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [x] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/#/how-to-open-a-pull-request).
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

Ideally https://github.com/freeCodeCamp/freeCodeCamp/pull/50099 should be merged first.

### CODE REVIEW

File: api/src/plugins/cors.ts [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+const cors: FastifyPluginCallback = (fastify, _options, done) => {
+  fastify.addHook('onRequest', async (req, reply) => {
+    const origin = req.headers.origin;
+    if (origin && allowedOrigins.includes(origin)) {
+      void reply.header('Access-Control-Allow-Origin', origin);
+    } else {
+      // TODO: Discuss if this is the correct approach. Standard practice is to
+      // reflect one of a list of allowed origins and handle development
+      // separately. If we switch to that approach we can replace use
+      // @fastify/cors instead.
```
Comment (naomi-lgbt) [inline api/src/plugins/cors.ts:26]: Should CORS not be mounted conditionally, such that it is disabled in development entirely?

Comment (ojeytonwilliams) [reply api/src/plugins/cors.ts:26]: There are a few things we can do, but we need to provide something for that header (either '*' or the actual origin). If we don't, any browser will throw CORS errors. At least, that's my understanding.

I think

> if development, then add Access-Control-Allow-Origin: origin

should be fine, but it's a departure from our current approach.

File: api/jest.utils.ts [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+declare global {
+  // eslint-disable-next-line no-var
+  var fastifyTestInstance: Awaited<ReturnType<typeof build>> | undefined;
+}
+
+export function superGet(endpoint: string): request.Test {
+  return request(fastifyTestInstance?.server)
+    .get(endpoint)
+    .set('Origin', 'https://www.freecodecamp.org');
+}
```
Comment (ShaunSHamilton) [inline api/jest.utils.ts:14]: When/why/how/where will we be running Jest in production?

Comment (ojeytonwilliams) [reply api/jest.utils.ts:14]: Never, but we can mock the FCC_NODE_ENV value.

It just bugged me that I couldn't test the behaviour I was interested in (i.e. production), so now we can.

I expect to refine/abandon this as we write more tests, though.

**Review by ShaunSHamilton — APPROVED** [review]
Does not break Gitpod 👍

**Review by Sembauke — APPROVED** [review]

**Pull request merged**